crates/dbsp/src/operator/input_upsert.rs

```diff
@@ -0,0 +1,344 @@
+use crate::{
+    algebra::{AddAssignByRef, HasOne, HasZero, PartialOrder, ZRingValue},
+    circuit::{
+        operator_traits::{BinaryOperator, Operator},
+        OwnershipPreference, Scope, WithClock,
+    },
+    operator::trace::{DelayedTraceId, TraceAppend, TraceBounds, TraceId, Z1Trace},
+    trace::{
+        consolidation::consolidate, cursor::Cursor, Batch, BatchReader, Builder, Filter, Spine,
+        Trace,
+    },
+    utils::VecExt,
+    Circuit, DBData, DBTimestamp, Stream, Timestamp,
+};
+use rkyv::{Archive, Deserialize, Serialize};
+use size_of::SizeOf;
+use std::{borrow::Cow, marker::PhantomData, ops::Neg};
+
+#[derive(
+    Clone,
+    Debug,
+    Default,
+    SizeOf,
+    PartialEq,
+    Eq,
+    Hash,
+    PartialOrd,
+    Ord,
+    Archive,
+    Serialize,
+    Deserialize,
+)]
+#[archive_attr(derive(Ord, Eq, PartialEq, PartialOrd))]
+#[archive(compare(PartialEq, PartialOrd))]
+pub enum Update<V: DBData, U: DBData> {
+    Insert(V),
+    #[default]
+    Delete,
+    Update(U),
+}
+
+pub type PatchFunc<V, U> = Box<dyn Fn(&mut V, U)>;
+
+impl<C, K, V, U> Stream<C, Vec<(K, Update<V, U>)>>
+where
+    K: DBData,
+    V: DBData,
+    U: DBData,
+    C: Circuit,
+    <C as WithClock>::Time: DBTimestamp,
+{
+    /// Convert an input stream of upserts into a stream of updates to a
+    /// relation.
+    ///
+    /// The input stream carries changes to a key/value map in the form of
+    /// _upserts_.  An upsert assigns a new value to a key (or deletes the key
+    /// from the map) without explicitly removing the old value, if any.  The
+    /// operator converts upserts into batches of updates, which is the input
+    /// format of most DBSP operators.
+    ///
+    /// The operator assumes that the input vector is sorted by key; however,
+    /// unlike the [`Stream::upsert`] operator it allows the vector to
+    /// contain multiple updates per key.  Updates are applied one by one in
+    /// order, and the output of the operator reflects cumulative effect of
+    /// the updates.  Additionally, unlike the [`Stream::upsert`] operator,
+    /// which only supports inserts, which overwrite the entire value with a
+    /// new value, and deletions, this operator also supports updates that
+    /// modify the contents of a value, e.g., overwriting some of its
+    /// fields.  Type argument `U` defined the formal of modifications,
+    /// and the `patch_func` function applies update of type `U` to a value of
+    /// type `V`.
+    ///
+    /// This is a stateful operator that internally maintains the trace of the
+    /// collection.
+    pub fn input_upsert<B>(&self, patch_func: PatchFunc<V, U>) -> Stream<C, B>
+    where
+        B::R: DBData + ZRingValue,
+        B: Batch<Key = K, Val = V, Time = ()>,
+    {
+        let circuit = self.circuit();
+
+        // We build the following circuit to implement the upsert semantics.
```

Comment:
Isn't this documentation important enough to be part of the doc generated for the operator?

Reply:
We have many of these diagrams in the code, and they are never in doc comments. Strictly speaking, these are implementation details. If we want to move them to doc comments, we should probably do this wholesale for all of them. Not in this PR though.